blueprints/to-locks.md

```diff
@@ -0,0 +1,279 @@
+    - A CDN notification displaying which CDN is locked by which user
+    - An "unlock" option that appears next to the notification, only for the user who has locked the CDN
+    - A way for the `admin` user to be able to unlock CDNs on other users' behalf
+    
+- Snapshot/ Queue Updates Page
+    - Disable the snap/ queue button if another user has the lock on that CDN.
+
+### Traffic Ops Impact
+
+`/cdn_locks`
+- List all CDN Locks
+- GET+POST+DELETE
+
+Traffic Ops will need to add the logic to check for locks before snapping/ queueing a CDN. It'll also need to account for
+`shared` vs `exclusive` locks, and forbid a user from snapping/ queueing a CDN if another user possesses a lock on that CDN.
+The following endpoints will need to handle the locks logic:
+- `/cachegroups/{{ID}}/queue_update`
+- `/cdns/{{ID}}/queue_update`
+- `/servers/{{hostname}}/queue_update`
+- `/snapshot`
+- `/topologies/{{name}}/queue_update`
+
+#### REST API Impact
+
+The following is the JSON representation of a `CDN_Lock` object:
+```JSON
+{
+  "username": "foo",
+  "cdn": "cdn1",
+  "message": "snapping cdn",
+  "shared": false,
+  "creator": true,
+  "lastUpdated": "2021-05-10T16:03:34-06:00"
+}
+```
+
+The following table describes the top level `CDN_Lock` object:
+
+| field       | type                        | optionality | description                                                                             |
+| ----------- | --------------------------- | ----------- | ----------------------------------------------------------------------------------------|
+| username    | string                      | required    | the user name of the user that wants to acquire a lock on the CDN                       |
+| cdn         | string                      | required    | the name of the CDN on which the lcok needs to be acquired                              |
+| message     | string                      | optional    | the message stating a reason behind locking the CDN                                     |
+| shared      | boolean                     | required    | whether or not this is a shared lock                                                    |
+| creator     | boolean                     | required    | whether or not the requesting `username` is the first one to acquire a lock on `cdn`| 
+| lastUpdated | time                        | required    | the last time this lock was updated                                                     |
+
+**API constraints:**
+- a `username` and `cdn` combination must be unique if `shared` is set to `false`. In other words, there can be only one `exclusive`
+lock for a CDN
+- a CDN can have multiple `shared` locks
+- the `creator` will be set to `true` if the user is the first one acquiring a lock for the specified `cdn`
+- a user can snap/ queue a CDN only if the `creator` field corresponding to the `username` and `cdn` combination is `true`
```

Comment:
a user can snap/queue a CDN only if their username matches the `creator` username?

Reply:
yes, you're correct. Because in exclusive locks, no one else has the lock, and in shared locks, only the creator/ owner of the lock is allowed to snap/ queue the cdn.